### PR TITLE
Store Services: Improve label request error messaging and options to retry

### LINKED ISSUE #1278: Enhancement/Bug: Better label print erroring
Steps to reproduce issue:
1. Create a product with no dimensions
2. Purchase that product
3. Go to print a label for that order
4. Result:
screen shot 2018-01-12 at 10 53 34 am
screen shot 2018-01-12 at 10 53 36 am

Logs:
> 01-11-2018 @ 15:48:34 - Error. Unable to get shipping rate(s) for wc_services_usps instance id 28. (calculate_shipping)
> 01-11-2018 @ 15:48:34 - product_missing_weight Product ( ID: 1404 ) did not include a weight. Shipping rates cannot be calculated. (calculate_shipping)

From the users perspective, they just see the Print button spinning endlessly so they aren't sure what's going on.  It would be great if we could error out here letting them know we can't print a label because there is no weight for a product.  

Ideally, purchasing a label wouldn't be possible without a weight to start with.

### PR DESCRIPTION
**Pull request by dechov**
- On a label rates request error, provide a clear way to retry, without having to return to "Packages" step or reopen dialog (message text is from the API client, and not part of this PR):

screen shot 2018-03-26 at 3 53 06 pm

- On label reprint request error, disable "busy" state on the action button, and show an error with a "Retry" option (fixes https://github.com/Automattic/woocommerce-services/issues/1278):

screen shot 2018-03-26 at 3 59 17 pm

- On label status fetch (checking for refund status from server) error, shows at most one error notice (instead of one per label), and adds context to error, since it isn't otherwise obvious what failed. First sentence is from the action, and the rest is from the API client:

screen shot 2018-03-26 at 3 51 55 pm

### CODE REVIEW

**Comment by kwight** [comment]
Hi @dechov , is this still active? If so, could you add some step-by-step instructions for testing please? Many of us are not familiar with Woo specifics but would love to help review from time to time!

**Review by DanReyLop — APPROVED** [review]
The `retry` buttons are a great touch!

Just think for like a few seconds about my comment and then you can :shipit: :)

**Comment by jeffstieler** [comment]
Ping @dechov - @DanReyLop approved.

**Comment by dechov** [comment]
@DanReyLop I think your comment exposes another issue with the way this change avoids multiple notices: there's a chance you'll run into an annoying behavior where you X out the notice, and it pops back up for another label status request (repeatedly). A better solution would be to handle `Promise.all` of the responses at once.

I'll look back at this in the next day or two to try and improve this, hopefully in a way that addresses your initial concern 👍

**Comment by dechov** [comment]
@DanReyLop Didn't manage to address your concern – even identifying the label(s) with the error would mean importing a selector, since we're using indices not derivable from the label data alone. Let me know if you think it's worth pursuing further.

Once I sort out the rebase on newly formatted files in master, will seek final approval, having modified the way the error notice is dispatched.

**Comment by DanReyLop** [comment]
> since we're using indices not derivable from the label data alone.

Oh that's right, nevermind then. 99% of the orders will have a single label or a batch of labels purchased at the same time, most of the time if they fail, they will fail for the same reason.

I'll give it a final pass once you rebase.

**Comment by dechov** [comment]
Just noticed another non-ideal case when a label rates request succeeds but subsequent ones fail:

previous-rates-not-reset

The loading indicator turns back into a "success" checkmark, yet the rates are inconsistent with the saved package state. We should probably reset the rates state whenever rates are fetched, right?

File: client/extensions/woocommerce/woocommerce-services/state/shipping-label/actions.js [hunk 1]
```diff
@@ -873,31 +873,33 @@ export const fetchLabelsStatus = ( orderId, siteId ) => ( dispatch, getState ) =
...
-					error,
-				} );
-				if ( error ) {
-					dispatch( NoticeActions.errorNotice(
-						`Failed to retrieve shipping label refund status. ${ error }`,
-						{ id: 'wcs-fetch-label-status' }
-					) );
-				}
+
+		const setIsSaving = () => {
```
Comment (DanReyLop) [inline client/extensions/woocommerce/woocommerce-services/state/shipping-label/actions.js]: This function name doesn't make sense anymore. What about `storeResponse`?

**Comment by DanReyLop** [comment]
> We should probably reset the rates state whenever rates are fetched, right?

Yes, that makes the most sense I think. Could you please add an issue for that?

**Review by DanReyLop — APPROVED** [review]
You need to rebase once again to get this change https://github.com/Automattic/wp-calypso/pull/25029

Once you get CircleCI to pass (and fix the one nitpick I had), :shipit:

File: client/extensions/woocommerce/woocommerce-services/state/shipping-label/actions.js (hunk 1 above)
Comment (dechov) [reply client/extensions/woocommerce/woocommerce-services/state/shipping-label/actions.js]: How about just having it anonymous, inline?

**Pull request merged**